README.md

```diff
@@ -201,6 +201,17 @@ For x86-64, the `vmlinux` kernel image will then be located at
 For AArch64, the `Image` kernel image will then be located at
 `linux-cloud-hypervisor/arch/arm64/boot/Image`.
 
+#### Handling bzImage
+
+On x86-64, `bzImage` is a common format for Linux kernels. Cloud Hypervisor does not support booting `bzImage` directly. To boot a `bzImage`, you either have to wrap it in a disk image (see below) or unpack the contained `vmlinux` file using a script from the Linux kernel source directory.
+
+The `vmlinux` can be unpacked from a `bzImage` using:
+
+```shell
+# In the Linux source tree
+$ ./scripts/extract-vmlinux bzImage > vmlinux
```

Comment:
Thank you for taking the time to improve our documentation - i'm not sure about this. If the kernel wasn't built in the correct way the PVH headers won't be present and thus the vmlinux file from the output might not work. It would be better to reopen the supporting bzImage topic / just have users use a firmware or direct kernel boot from a kernel with the right PVH setup.

Reply:
I can also improve the error message to be actionable in case the PVH header is not present. This would then guide users to what they need to do.

Btw, is there a list of required kernel options? Reverse engineering it from the provided config is a bit cumbersome.

As for the bzImage topic: Where is the best place to discuss this?

Reply:
I did some code archaeology and identified when/why the bzImage support was removed. We decided to remove the support as part of the removal of the Linux Boot (plain vmlinux) support. This was because it has been superseded by the use of the PVH which is an officially supported method for direct kernel boot. As using bzImage for this purpose is not officially supported I do not want to support that in CH. 

Further since the conversion of arbitrary bzImages might not produce binaries that can be booted (and create more support requests/maintenance burden) I do not want to include the documentation change you made.

If users want to kernel direct boot they should use the kernel config we advise and build it in the way we document. That is the only supported method. (I know e.g. Kata Containers has their own kernel but they also directly support their users so it doesn't create extra work here.)

Reply:
@rbradford 

I understand the complex topic of keeping maintenance low. But I think it's not that easy. From my perspective, it feels like the user needs to deal with more moving parts than they really need to.

The firmware path is unfortunately not a drop-in replacement and requires that the user builds a disk image. We ended up at the direct kernel boot, because rust-hypervisor-firmware is not complete enough to boot a recent systemd-boot (I'll open an issue when I get to it). Since we actually don't want to boot from disk images, our colleague tried to boot the kernel directly and failed due to missing information in the README.

Given this sentence in the README:

> provided the required options are enabled any recent kernel will suffice.

I would love to be able to tell users that they can just grab the kernel from their distro and get started. I checked NixOS and Fedora and they have `CONFIG_PVH` enabled. The NixOS kernel also just works (minus the bzImage/vmlinux issue).

So for the path forward, I see a couple of options to make things easier for users:

1. Support booting from bzImage. Autodetect whether users hand in a PVH kernel or a bzImage.
  - This seems pretty straightforward. All the pieces are there. It would reduce the support burden, because pretty much all kernels Just Work™.
2. Implement direct kernel boot in rust-hypervisor-firmware, so you don't need to build a disk image.
  - This could work somewhat like direct kernel boot in Qemu. The firmware gets everything via fw-cfg.
 
Either way would have prevented us from popping up here. :) I'm going to look into 1. for now as exercise, even though it might not be merged. I just want to be more familiar with the code base. :)

> As using bzImage for this purpose is not officially supported

Can you elaborate?